middleware.go

```diff
@@ -0,0 +1,28 @@
+package mux
+
+import "net/http"
+
+// MiddlewareFunc is a function which receives an http.Handler and returns another http.Handler.
+// Typically, the returned handler is a closure which does something with the http.ResponseWriter and http.Request passed
+// to it, and then calls the handler passed as parameter to the MiddlewareFunc.
+type MiddlewareFunc func(http.Handler) http.Handler
+
+// middleware interface is anything which implements a MiddlewareFunc named Middleware.
+type middleware interface {
+	Middleware(handler http.Handler) http.Handler
+}
+
+// MiddlewareFunc also implements the Middleware interface.
+func (mw MiddlewareFunc) Middleware(handler http.Handler) http.Handler {
+	return mw(handler)
+}
+
+// AddMiddlewareFunc appends a MiddlewareFunc to the chain.
+func (r *Router) AddMiddlewareFunc(mwf MiddlewareFunc) {
```

Comment:
`AddMiddleware`, `Add` or `UseMiddleware` or just `Use` might be less verbose here.

* `chi` and `echo` call their methods `Use`
* [express](http://expressjs.com/en/guide/using-middleware.html) also calls it `use`
* Downside to `Use` - not entirely self-documenting.

I'd be OK with just calling this function `Middleware` as "Add" is obvious, and subtracting it/removing it later is (mostly) a non-option.

Reply:
I don't like the `AddMiddlewareFunc` name either, seems way too long to me. But I think we should keep the "func" part somewhere, as it makes what this function expects very clear, and avoids the need for a breaking change if the `middleware` interface is exported in the future.

What about `UseFunc`?

Reply:
Let’s just stick with `Use` - as that’s familiar and aligns with the `Handle` (takes a http.Handler) and `HandleFunc` (takes a HandlerFunc) on `Router`. This somewhat sits in the middle, but familiarity wins out in lieu of other objections.

Reply:
Done. I've renamed the internal `addMiddleware` to `useInterface`, to avoid collision with `Use`.